core/src/test/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactoryTest.java

```diff
@@ -106,6 +106,56 @@ public void testSeconds() throws NumericException, SqlException {
         }
     }
 
+    @Test
+    public void testMillis() throws NumericException, SqlException {
+        StringSink sink = new StringSink();
+        for (int k = 0; k < 1001; k++) {
+            sink.clear();
+            if (k > 0) {
+                sink.put(k).put('T');
+            } else {
+                sink.put('T');
+            }
+            TimestampSampler sampler = TimestampSamplerFactory.getInstance(sink, 120);
+            Assert.assertNotNull(sampler);
+
+            final long n = Timestamps.MILLI_MICROS * (k == 0 ? 1 : k);
+            long timestamp = TimestampFormatUtils.parseUTCTimestamp("2018-04-15T10:23:00.000000Z");
+            timestamp = timestamp - timestamp % n;
+            for (int i = 0; i < n; i += 4) {
+                long actual = sampler.round(timestamp + i);
+                if (timestamp != actual) {
+                    Assert.fail("Failed at: " + sink + ". Expected: " + Timestamps.toString(timestamp) + ", actual: " + Timestamps.toString(actual));
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testMicros() throws NumericException, SqlException {
+        StringSink sink = new StringSink();
+        for (int k = 0; k < 1001; k++) {
```

Comment:
Thank you @KevinChen16 looking much closer to a merge, although there is now a lot of repeated code which would be nice to refactor so that the tests become more readable. 

Given this new method (please add it):

```
private static TimestampSampler createTimestampSampler(int bucketSize, char units, StringSink sink) throws SqlException {
        sink.clear();
        if (bucketSize > 0) {
            sink.put(bucketSize);
        }
        sink.put(units);
        TimestampSampler sampler = TimestampSamplerFactory.getInstance(sink, 120);
        Assert.assertNotNull(sampler);
        return sampler;
    }
```

would you please refactor TSFT to use it?. Then, for `testMicros()` you are only testing `k` up to 1001, when the micros range is up to 1000001, that is why I suggested trying k values for up to 1000 attempts, where at each attempt k is randomly chosen like this:

```
@Test
    public void testMicros() throws NumericException, SqlException {
        final StringSink sink = new StringSink();
        final long ts = TimestampFormatUtils.parseUTCTimestamp("2022-04-23T10:33:00.123456Z");
        final Rnd rand = new Rnd();
        for (int j = 0; j < 1000; j++) {
            final int k = rand.nextInt(1000001);
            final TimestampSampler sampler = createTimestampSampler(k, 'U', sink);
            final long bucketSize = (k == 0 ? 1 : k);
            final long expectedTs = ts - ts % bucketSize;
            for (int i = 0; i < bucketSize; i+=4) {
                long actualTs = sampler.round(expectedTs + i);
                if (expectedTs != actualTs) {
                    Assert.fail(String.format(
                            "Failed at: %s, i: %d. Expected: %s, actual: %s",
                            sink, i, Timestamps.toString(expectedTs), Timestamps.toString(actualTs))
                    );
                }
            }
        }
    }
```
Could you please amend.

Thank you in advance!!!

Reply:
Sure! I'll upload it soon! Thank you for your advice!